cli_tools/common/utils/logging/firelog_logger.go

```diff
@@ -0,0 +1,311 @@
+		ToolAction:   l.Action,
+		Status:       statusSuccess,
+		InputParams:  &l.Params,
+		OutputInfo:   getOutputInfo(w, nil),
+	}
+
+	return logEvent, l.sendLogToServer(logEvent)
+}
+
+// logFailure logs a "failure" info to server
+func (l *FirelogLogger) logFailure(err error, w *daisy.Workflow) (*ComputeImageToolsLogExtension, logResult) {
+	logEvent := &ComputeImageToolsLogExtension{
+		ID:           l.ID,
+		CloudBuildID: os.Getenv(daisyutils.BuildIDOSEnv),
+		ToolAction:   l.Action,
+		Status:       statusFailure,
+		InputParams:  &l.Params,
+		OutputInfo:   getOutputInfo(w, err),
+	}
+
+	return logEvent, l.sendLogToServer(logEvent)
+}
+
+func getFailureReason(err error) string {
+	return daisyutils.RemovePrivacyLogTag(err.Error())
+}
+
+func getAnonymizedFailureReason(err error) string {
+	derr := daisy.ToDError(err)
+	if derr == nil {
+		return ""
+	}
+	anonymizedErrs := []string{}
+	for _, m := range derr.AnonymizedErrs() {
+		anonymizedErrs = append(anonymizedErrs, daisyutils.RemovePrivacyLogInfo(m))
+	}
+	return strings.Join(anonymizedErrs, "\n")
+}
+
+func getOutputInfo(w *daisy.Workflow, err error) *OutputInfo {
+	o := OutputInfo{
+		TargetSizeGb: getInt64Value(w.GetSerialConsoleOutputValue(targetSizeGb)),
+		SourceSizeGb: getInt64Value(w.GetSerialConsoleOutputValue(sourceSizeGb)),
+	}
+
+	if err != nil {
+		o.FailureMessage = getFailureReason(err)
+		o.FailureMessageWithoutPrivacyInfo = getAnonymizedFailureReason(err)
+	}
+
+	return &o
+}
+
+func getInt64Value(s string) int64 {
+	i, err := strconv.ParseInt(s, 0, 64)
+	if err != nil {
+		return 0
+	}
+	return i
+}
+
+func (l *FirelogLogger) runWithServerLogging(function func() (*daisy.Workflow, error)) *ComputeImageToolsLogExtension {
+	var event *ComputeImageToolsLogExtension
+
+	// Send log asynchronously. No need to interrupt the main flow when failed to send log, just
+	// keep moving.
+	var wg sync.WaitGroup
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		l.logStart()
+	}()
+
+	if w, err := function(); err != nil {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			event, _ = l.logFailure(err, w)
+			log.Println(event.OutputInfo.FailureMessage)
+		}()
+	} else {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			event, _ = l.logSuccess(w)
+		}()
+	}
+
+	wg.Wait()
+	return event
+}
+
+// RunWithServerLogging runs the function with server logging
+func RunWithServerLogging(action string, params InputParams, function func() (*daisy.Workflow, error)) {
+	l := NewFirelogLogger(action, params)
+	l.runWithServerLogging(function)
+}
+
+func (l *FirelogLogger) sendLogToServer(logEvent *ComputeImageToolsLogExtension) logResult {
+	return l.sendLogToServerWithRetry(logEvent, 3)
+}
+
+func (l *FirelogLogger) sendLogToServerWithRetry(logEvent *ComputeImageToolsLogExtension, maxRetry int) logResult {
+	logMutex.Lock()
+	defer logMutex.Unlock()
+
+	for i := 0; i < maxRetry; i++ {
+		// Before sending a new request, wait for a while if server asked to do so
+		if nextRequestWaitMillis > 0 {
```

Comment:
Can this tool be run many times in quick succession? I think I'd be more comfortable if this value was serialized to disk and read from there. Keeping it in memory isn't likely to achieve the flow control we want.

Reply:
Yes, this tool can be run many times. The problem is that it is mostly executed in docker, so it;s in an isolated environment.